test/new-e2e/tests/containers/base_test.go

```diff
@@ -11,40 +11,94 @@ import (
 	"strings"
 	"time"
 
+	"github.com/samber/lo"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/suite"
 
+	"github.com/DataDog/agent-payload/v5/gogen"
 	"github.com/DataDog/datadog-agent/test/fakeintake/aggregator"
 	fakeintake "github.com/DataDog/datadog-agent/test/fakeintake/client"
 )
 
 type baseSuite struct {
 	suite.Suite
 
+	startTime  time.Time
+	endTime    time.Time
 	Fakeintake *fakeintake.Client
 }
 
-func (suite *baseSuite) testMetric(metricName string, filterTags []string, expectedTags []*regexp.Regexp) {
-	suite.Run(fmt.Sprintf("%s{%s}", metricName, strings.Join(filterTags, ",")), func() {
+func (suite *baseSuite) SetupSuite() {
+	suite.startTime = time.Now()
+}
+
+func (suite *baseSuite) TearDownSuite() {
+	suite.endTime = time.Now()
+}
+
+type testMetricArgs struct {
+	filter testMetricFilterArgs
+	expect testMetricExpectArgs
+}
+
+type testMetricFilterArgs struct {
+	name string
+	tags []string
+}
+
+type testMetricExpectArgs struct {
+	tags  *[]string
+	value *testMetricExpectValueArgs
+}
+
+type testMetricExpectValueArgs struct {
+	min float64
+	max float64
+}
+
+func (suite *baseSuite) testMetric(args *testMetricArgs) {
+	suite.Run(fmt.Sprintf("%s{%s}", args.filter.name, strings.Join(args.filter.tags, ",")), func() {
+		var expectedTags []*regexp.Regexp
+		if args.expect.tags != nil {
+			expectedTags = lo.Map(*args.expect.tags, func(tag string, _ int) *regexp.Regexp { return regexp.MustCompile(tag) })
+		}
+
 		suite.EventuallyWithTf(func(collect *assert.CollectT) {
 			metrics, err := suite.Fakeintake.FilterMetrics(
-				metricName,
-				fakeintake.WithTags[*aggregator.MetricSeries](filterTags),
+				args.filter.name,
+				fakeintake.WithTags[*aggregator.MetricSeries](args.filter.tags),
 			)
 			if err != nil {
 				collect.Errorf("%w", err)
 				return
 			}
 			if len(metrics) == 0 {
-				collect.Errorf("No `%s{%s}` metrics yet", metricName, strings.Join(filterTags, ","))
+				collect.Errorf("No `%s{%s}` metrics yet", args.filter.name, strings.Join(args.filter.tags, ","))
 				return
 			}
 
 			// Check tags
-			if err := assertTags(metrics[len(metrics)-1].GetTags(), expectedTags); err != nil {
-				collect.Errorf("Tags mismatch on `%s`: %w", metricName, err)
-				return
+			if expectedTags != nil {
+				if err := assertTags(metrics[len(metrics)-1].GetTags(), expectedTags); err != nil {
+					collect.Errorf("Tags mismatch on `%s`: %w", args.filter.name, err)
+				}
+			}
+
+			// Check value
+			if args.expect.value != nil {
+				if lo.CountBy(metrics[len(metrics)-1].GetPoints(), func(v *gogen.MetricPayload_MetricPoint) bool {
+					return v.GetValue() >= args.expect.value.min &&
+						v.GetValue() <= args.expect.value.max
+				}) == 0 {
+					collect.Errorf(
+						"No value of `%s{%s}` is in the range [%f;%f]",
+						args.filter.name,
+						strings.Join(args.filter.tags, ","),
+						args.expect.value.min,
+						args.expect.value.max,
```

Comment:
Good remark !

I could indeed use `WithMetricValueInRange` to filter the metrics returned by the fake intake client API.
In a similar way, if `WithTags` was supporting regular expressions, I could have used `WithTags` to filter metrics based on the expected tags.
(I need regular expressions because some tags generated by the agent are containing identifiers that cannot be predicted. So, the whole value of some tags cannot be hard-coded in the tests. Yet, I want to validate that a tag with a given key exists and that the value respects a given format.)

This would be sufficient to tell if a test is successful or not.
If the test managed to get metrics with the value and tags filtered, it’s a success, else, it’s a failure.

The issue is that, if `filterMetrics(metricName, WithTags(…), WithMetricValueInRange(…))` returns nothing, I can make the test fail. Fine. But the only error message I could produce to the user would be `Failed to find metrics metricName with tags … and a value in range …`. I wouldn’t be able to provide any more information about exactly what was missing.

With my `testMetric` approach,
* First, pass the *very minimum* filtering options to the fake intake client to guarantee that we retrieve only the series we’re interested in. And then
* Look for the tags and values on the returned series,

it allows me to generate error messages telling:
* if it’s the whole metric that is completely missing or
* if it’s only a subset of the expected tags that is missing or
* if it’s only the value which isn’t in the expected range.

It is extremely important as those different errors are the symptoms of failures of different components.

* If I completely miss the series, it might be because the auto-discovery didn’t schedule the check.
* If I only miss a subset of the tags, it might be because only a specific collector of the tagger is misbehaving.
* If it’s only the value which is wrong, it means that the check has been properly scheduled, that the tagger is working but maybe there’s a bug in the check itself.

Here are some examples:

https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/356564951
```
=== Failed
=== FAIL: tests/containers TestKindSuite/TestRedis/kubernetes_state.deployment.replicas_available{kube_deployment:redis,kube_namespace:workload-redis} (120.00s)
    base_test.go:29: No `kubernetes_state.deployment.replicas_available{kube_deployment:redis,kube_namespace:workload-redis}` metrics yet
```
Here, the whole metric is missing.
So, it can be the redis deployment which wasn’t started.

https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/354548669
```
=== Failed
=== FAIL: tests/containers TestECSSuite/TestNginx/nginx.net.request_per_s{} (120.00s)
    base_test.go:29: Tags mismatch on `nginx.net.request_per_s`: missing tags: ^cluster_name:ci-22199720-4670-ecs-cluster-ecs$, ^task_version:[[:digit:]]+$, ^task_name:.*-nginx-ec2$, ^task_family:.*-nginx-ec2$, ^ecs_cluster_name:ci-22199720-4670-ecs-cluster-ecs$, ^ecs_container_name:nginx$, ^task_arn:
```
Here, what is missing is only the subset of tags that are coming from the ECS collector.
So, I would focus of the tagger.

If all the filtering had been done in the fake intake client, I wouldn’t have been able to provide that level of details in the error messages.

Reply:
Thank you for this detailed analysis ! This is very much related to this ticket around returning a `MatchResult` rather than a list of metrics, possibly empty 

https://datadoghq.atlassian.net/browse/APL-2201

I can see how the test domain knows what is wrong with each unmatched requirement, we could maybe have some primitives that can be reused. I agree that today `Filter*` on fakeintake client side hides too much info